File: src/render/transitions/trim-mixed-concat.ts
import path from 'node:path';
import { execFileSync } from 'node:child_process';
import { RenderError } from '../../errors.js';
import { updateProgress } from '../task.js';
import type { QualityPreset, ResolutionPreset } from '../../presets.js';
import type { RenderTask } from '../task.js';
import type { VideoClip } from '../types.js';
import {
  collectClips,
  normalizeClips,
  type TransitionHandlerArgs,
  type TransitionHandlerResult,
} from './shared.js';

export type JunctionType = 'flash-black' | 'dissolve' | 'cut';

export function parseJunctionType(raw: unknown): JunctionType {
  if (raw === 'flash-black' || raw === 'dissolve' || raw === 'cut') return raw;
  return 'cut';
}

/**
 * 混合转场 filter_complex 构建：
 *
 * 每个 clip 拆成三部分（视情况有或无）：
 *   head  — 前段 half_D 的淡入段（仅当上一个 junction 是 flash-black）
 *   body  — 主体段（head..tail 之间）
 *   tail  — 末尾转场段（取决于下一个 junction 类型）
 *
 * 连接点段：
 *   flash-black → [fb_tail_i]  [fb_head_{i+1}]  （两段独立，依次拼接）
 *   dissolve    → [diss_trans_i]                 （前段尾叠后段首帧，overlay）
 *   cut         → （无额外段）
 */
export function ffmpegTrimMixedConcat(
  ffmpegPath: string,
  clips: VideoClip[],
  junctions: JunctionType[],
  trimStart: number,
  outputPath: string,
  qualPreset: QualityPreset,
  task: RenderTask,
  resPreset: ResolutionPreset,
  transitionDuration: number,
): void {
  const n = clips.length;
  const D = transitionDuration;
  const halfD = D / 2;
  const fps = resPreset.fps;
  const frameDuration = 1 / fps;
  const minSeg = frameDuration / 2;
  const fmt = (v: number) => v.toFixed(6);

  // 每个 input 加 -ss trimStart 跳过开头
  const inputArgs: string[] = [];
  for (const clip of clips) {
    if (trimStart > 0) inputArgs.push('-ss', String(trimStart));
    inputArgs.push('-i', clip.src);
  }

  // trim 后的有效时长
  const durations = clips.map((c) => Math.max(0, c.duration - trimStart));

  // 单段直通
  if (n === 1) {
    execFileSync(
      ffmpegPath,
      [
        ...inputArgs,
        '-filter_complex', `[0:v]setpts=PTS-STARTPTS,fps=fps=${fps}[vout]`,
        '-map', '[vout]',
        '-c:v', 'libx264',
        '-preset', qualPreset.ffmpegPreset,
        '-crf', String(qualPreset.crf),
        '-pix_fmt', 'yuv420p',
        '-an',
        '-y', outputPath,
      ],
      { timeout: 600_000 },
    );
    updateProgress(task, 1);
    return;
  }

  // 计算每段首尾被转场占用的时长
  const tailConsumed: number[] = new Array(n).fill(0); // 末尾被出口转场占用
  const headConsumed: number[] = new Array(n).fill(0); // 开头被入口转场占用

  for (let j = 0; j < junctions.length; j++) {
    const t = junctions[j];
    const dj = durations[j];
    const djn = durations[j + 1];

    if (t === 'flash-black') {
      const eff = Math.min(halfD, dj / 2, djn / 2);
      if (eff < halfD - 0.001) {
        console.warn(
          `  ⚠ 连接点 ${j + 1} (flash-black) 附近片段时长不足，闪黑已收缩至 ${(eff * 2).toFixed(2)}s`,
        );
      }
      tailConsumed[j] = eff;
      headConsumed[j + 1] = eff;
    } else if (t === 'dissolve') {
      const eff = Math.min(D, dj / 2);
      if (eff < D - 0.001) {
        console.warn(
          `  ⚠ 连接点 ${j + 1} (dissolve) 片段 ${clips[j].key} 时长不足，叠化已收缩至 ${eff.toFixed(2)}s`,
        );
      }
      tailConsumed[j] = eff;
      // dissolve 不占用下一段的开头
    }
    // cut: 不占用任何时长
  }

  const filterParts: string[] = [];
  const segmentLabels: string[] = [];

  for (let i = 0; i < n; i++) {
    const d = durations[i];
    const bodyStart = headConsumed[i];
    const bodyDuration = d - headConsumed[i] - tailConsumed[i];

    // 主体段
    if (bodyDuration > minSeg) {
      filterParts.push(
        `[${i}:v]trim=start=${fmt(bodyStart)}:duration=${fmt(bodyDuration)},setpts=PTS-STARTPTS,fps=fps=${fps}[body${i}]`,
      );
      segmentLabels.push(`[body${i}]`);
    } else if (bodyDuration > 0) {
      console.warn(`  ⚠ 片段 ${clips[i].key} 主体过短（${bodyDuration.toFixed(3)}s），已跳过`);
    }

    // 出口转场段（在 body 之后）
    if (i < n - 1) {
      const t = junctions[i];

      if (t === 'flash-black') {
        const eff = tailConsumed[i];

        // 当前段尾部：淡出至黑
        filterParts.push(
          `[${i}:v]trim=start=${fmt(d - eff)}:duration=${fmt(eff)},setpts=PTS-STARTPTS,` +
          `fade=t=out:st=0:d=${fmt(eff)},fps=fps=${fps}[fb_tail${i}]`,
        );
        segmentLabels.push(`[fb_tail${i}]`);

        // 下一段头部：从黑淡入
        filterParts.push(
          `[${i + 1}:v]trim=duration=${fmt(eff)},setpts=PTS-STARTPTS,` +
          `fade=t=in:st=0:d=${fmt(eff)},fps=fps=${fps}[fb_head${i + 1}]`,
        );
        segmentLabels.push(`[fb_head${i + 1}]`);

      } else if (t === 'dissolve') {
        const eff = tailConsumed[i];
        const stillDuration = Math.max(0, eff - frameDuration);

        // 当前段尾部：alpha 淡出
        filterParts.push(
          `[${i}:v]trim=start=${fmt(d - eff)}:duration=${fmt(eff)},setpts=PTS-STARTPTS,` +
          `format=rgba,fade=t=out:st=0:d=${fmt(eff)}:alpha=1,fps=fps=${fps}[diss_tail${i}]`,
        );

        // 下一段首帧静止垫底
        filterParts.push(
          `[${i + 1}:v]trim=end_frame=1,setpts=PTS-STARTPTS,` +
          `tpad=stop_mode=clone:stop_duration=${fmt(stillDuration)},fps=fps=${fps}[diss_still${i}]`,
        );

        // 叠化：静帧在下，淡出段在上
        filterParts.push(
          `[diss_still${i}][diss_tail${i}]overlay=eof_action=pass:shortest=1,format=yuv420p,fps=fps=${fps}[diss_trans${i}]`,
        );
        segmentLabels.push(`[diss_trans${i}]`);
      }
      // cut: 无额外段
    }
  }

  if (segmentLabels.length === 0) {
    throw new RenderError('没有可用的视频段（所有片段时长过短）');
  }

  filterParts.push(
    `${segmentLabels.join('')}concat=n=${segmentLabels.length}:v=1:a=0[vout]`,
  );

  const filterComplex = filterParts.join(';');
  const junctionDesc = junctions.join(' → ');

  console.log(
    `\n[2/3] FFmpeg 混合转场拼接 ${n} 段（剪头 ${trimStart}s，转场: ${junctionDesc}）...`,
  );

  execFileSync(
    ffmpegPath,
    [
      ...inputArgs,
      '-filter_complex', filterComplex,
      '-map', '[vout]',
      '-c:v', 'libx264',
      '-preset', qualPreset.ffmpegPreset,
      '-crf', String(qualPreset.crf),
      '-pix_fmt', 'yuv420p',
      '-an',
      '-y', outputPath,
    ],
    { timeout: 600_000 },
  );

  updateProgress(task, 1);
}

export function handleTrimMixedConcat(
  args: TransitionHandlerArgs,
): TransitionHandlerResult {
  const {
    rootDir, ffmpegPath, ffprobePath, request, videoInfo,
    qualPreset, resPreset, task, outDir, outFile,
  } = args;

  void ffprobePath;

  console.log(`\n[1/3] 收集并校验视频片段...`);

  const trimStart = typeof request.variables.trim_start === 'number'
    ? request.variables.trim_start
    : 2;

  const transitionDuration = typeof request.variables.transition_duration === 'number'
    ? request.variables.transition_duration
    : 0.5;

  const clips = collectClips(request.variables, request.templateInfo, videoInfo);
  if (clips.length === 0) {
    throw new RenderError('没有可用的视频片段（未提供任何视频）');
  }

  // 按连接点顺序读取转场类型（transition_1 = clip1→clip2，以此类推）
  const junctions: JunctionType[] = [];
  for (let j = 1; j < clips.length; j++) {
    junctions.push(parseJunctionType(request.variables[`transition_${j}`]));
  }

  console.log(
    `  共 ${clips.length} 个片段，剪头 ${trimStart}s，转场: [${junctions.join(', ')}]，时长 ${transitionDuration}s`,
  );

  const normalized = normalizeClips(rootDir, ffmpegPath, clips, qualPreset, resPreset);

  const outputPath = path.join(outDir, outFile);
  ffmpegTrimMixedConcat(
    ffmpegPath,
    normalized.clips,
    junctions,
    trimStart,
    outputPath,
    qualPreset,
    task,
    resPreset,
    transitionDuration,
  );

  return {
    cleanup: normalized.cleanup,
    outputPath,
  };
}

File: videocut/render/transitions/trim_mixed_concat.py
from __future__ import annotations

from pathlib import Path
from typing import Literal

from videocut.errors import RenderError
from videocut.presets import QualityPreset, ResolutionPreset
from videocut.render.task import RenderTask, update_progress
from videocut.render.transitions.shared import (
    TransitionHandlerArgs,
    TransitionHandlerResult,
    _run_ffmpeg,
    collect_clips,
    normalize_clips,
)
from videocut.render.types import VideoClip

JunctionType = Literal["flash-black", "dissolve", "cut"]


def parse_junction_type(raw: object) -> JunctionType:
    if raw in {"flash-black", "dissolve", "cut"}:
        return raw  # type: ignore[return-value]
    return "cut"


def ffmpeg_trim_mixed_concat(
    ffmpeg_path: str,
    clips: list[VideoClip],
    junctions: list[JunctionType],
    trim_start: float,
    output_path: str,
    qual_preset: QualityPreset,
    task: RenderTask,
    res_preset: ResolutionPreset,
    transition_duration: float,
) -> None:
    clip_count = len(clips)
    total_transition = transition_duration
    half_duration = total_transition / 2
    fps = res_preset.fps
    frame_duration = 1 / fps
    min_segment = frame_duration / 2
    fmt = lambda value: f"{value:.6f}"

    input_args: list[str] = []
    for clip in clips:
        if trim_start > 0:
            input_args.extend(["-ss", str(trim_start)])
        input_args.extend(["-i", clip.src])

    durations = [max(0.0, clip.duration - trim_start) for clip in clips]

    if clip_count == 1:
        _run_ffmpeg(
            [
                ffmpeg_path,
                *input_args,
                "-filter_complex",
                f"[0:v]setpts=PTS-STARTPTS,fps=fps={fps}[vout]",
                "-map",
                "[vout]",
                "-c:v",
                "libx264",
                "-preset",
                qual_preset.ffmpeg_preset,
                "-crf",
                str(qual_preset.crf),
                "-pix_fmt",
                "yuv420p",
                "-an",
                "-y",
                output_path,
            ],
            timeout=600,
        )
        update_progress(task, 1.0)
        return

    tail_consumed = [0.0] * clip_count
    head_consumed = [0.0] * clip_count
    for index, junction in enumerate(junctions):
        current_duration = durations[index]
        next_duration = durations[index + 1]
        if junction == "flash-black":
            effective = min(half_duration, current_duration / 2, next_duration / 2)
            if effective < half_duration - 0.001:
                print(
                    f"  Warning: junction {index + 1} flash-black shortened to {(effective * 2):.2f}s"
                )
            tail_consumed[index] = effective
            head_consumed[index + 1] = effective
        elif junction == "dissolve":
            effective = min(total_transition, current_duration / 2)
            if effective < total_transition - 0.001:
                print(f"  Warning: junction {index + 1} dissolve shortened to {effective:.2f}s")
            tail_consumed[index] = effective

    filter_parts: list[str] = []
    segment_labels: list[str] = []
    for index in range(clip_count):
        duration = durations[index]
        body_start = head_consumed[index]
        body_duration = duration - head_consumed[index] - tail_consumed[index]
        if body_duration > min_segment:
            filter_parts.append(
                f"[{index}:v]trim=start={fmt(body_start)}:duration={fmt(body_duration)},"
                f"setpts=PTS-STARTPTS,fps=fps={fps}[body{index}]"
            )
            segment_labels.append(f"[body{index}]")
        elif body_duration > 0:
            print(f"  Warning: {clips[index].key} body segment too short ({body_duration:.3f}s), skip")

        if index >= clip_count - 1:
            continue
        junction = junctions[index]
        if junction == "flash-black":
            effective = tail_consumed[index]
            filter_parts.append(
                f"[{index}:v]trim=start={fmt(duration - effective)}:duration={fmt(effective)},"
                f"setpts=PTS-STARTPTS,fade=t=out:st=0:d={fmt(effective)},fps=fps={fps}[fb_tail{index}]"
            )
            segment_labels.append(f"[fb_tail{index}]")
            filter_parts.append(
                f"[{index + 1}:v]trim=duration={fmt(effective)},setpts=PTS-STARTPTS,"
                f"fade=t=in:st=0:d={fmt(effective)},fps=fps={fps}[fb_head{index + 1}]"
            )
            segment_labels.append(f"[fb_head{index + 1}]")
        elif junction == "dissolve":
            effective = tail_consumed[index]
            still_duration = max(0.0, effective - frame_duration)
            filter_parts.append(
                f"[{index}:v]trim=start={fmt(duration - effective)}:duration={fmt(effective)},"
                f"setpts=PTS-STARTPTS,format=rgba,fade=t=out:st=0:d={fmt(effective)}:alpha=1,"
                f"fps=fps={fps}[diss_tail{index}]"
            )
            filter_parts.append(
                f"[{index + 1}:v]trim=end_frame=1,setpts=PTS-STARTPTS,"
                f"tpad=stop_mode=clone:stop_duration={fmt(still_duration)},"
                f"fps=fps={fps}[diss_still{index}]"
            )
            filter_parts.append(
                f"[diss_still{index}][diss_tail{index}]overlay=eof_action=pass:shortest=1,"
                f"format=yuv420p,fps=fps={fps}[diss_trans{index}]"
            )
            segment_labels.append(f"[diss_trans{index}]")

    if not segment_labels:
        raise RenderError("No usable video segments after transition building.")

    filter_parts.append(f"{''.join(segment_labels)}concat=n={len(segment_labels)}:v=1:a=0[vout]")
    print(
        f"\n[2/3] FFmpeg mixed-transition concat {clip_count} clips "
        f"(trim {trim_start}s, junctions {' -> '.join(junctions)})..."
    )
    _run_ffmpeg(
        [
            ffmpeg_path,
            *input_args,
            "-filter_complex",
            ";".join(filter_parts),
            "-map",
            "[vout]",
            "-c:v",
            "libx264",
            "-preset",
            qual_preset.ffmpeg_preset,
            "-crf",
            str(qual_preset.crf),
            "-pix_fmt",
            "yuv420p",
            "-an",
            "-y",
            output_path,
        ],
        timeout=600,
    )
    update_progress(task, 1.0)


def handle_trim_mixed_concat(args: TransitionHandlerArgs) -> TransitionHandlerResult:
    print("\n[1/3] Collecting input clips...")
    trim_start = (
        float(args.request.variables["trim_start"])
        if isinstance(args.request.variables.get("trim_start"), (int, float))
        else 2.0
    )
    transition_duration = (
        float(args.request.variables["transition_duration"])
        if isinstance(args.request.variables.get("transition_duration"), (int, float))
        else 0.5
    )
    clips = collect_clips(args.request.variables, args.request.template_info, args.video_info)
    if not clips:
        raise RenderError("No usable video clips were provided.")

    junctions = [parse_junction_type(args.request.variables.get(f"transition_{index}")) for index in range(1, len(clips))]
    print(
        f"  {len(clips)} clip(s), trim {trim_start}s, transitions {junctions}, duration {transition_duration}s"
    )
    normalized_clips, cleanup = normalize_clips(
        args.root_dir,
        args.ffmpeg_path,
        clips,
        args.qual_preset,
        args.res_preset,
    )
    output_path = str(Path(args.out_dir) / args.out_file)
    ffmpeg_trim_mixed_concat(
        args.ffmpeg_path,
        normalized_clips,
        junctions,
        trim_start,
        output_path,
        args.qual_preset,
        args.task,
        args.res_preset,
        transition_duration,
    )
    return TransitionHandlerResult(cleanup=cleanup, output_path=output_path)

